doc/reference/index.rst

```diff
@@ -23,3 +23,214 @@ API Reference
+sleep has no use for no-wait support and is implicitly **isr-ok**.
+
+.. _api_term_reschedule:
+
+reschedule
+==========
+
+The reschedule attribute is used on a function that can reach a
+:ref:`reschedule point <scheduling_v2>` within its execution.
+
+.. note::
+
+   The significance of this attribute is that when a rescheduling
+   function is invoked by a thread it is possible for that thread to be
+   suspended as a consequence of a higher-priority thread being made
+   ready.  Whether the suspension actually occurs depends on the
+   operation associated with the reschedule point and the relative
+   priorities of the invoking thread and the head of the ready queue.
+
+   Note that in the case of timeslicing, or reschedule points executed
+   from interrupts, any thread may be suspended in any function.
+
+   Functions that are not **reschedule** may be invoked from either
+   thread or interrupt context.
+
+   Functions that are **reschedule** may be invoked from thread context.
+
+   Functions that are **reschedule** but not **sleep** may be invoked
+   from interrupt context.
+
+.. _api_term_sleep:
+
+sleep
+=====
+
+The sleep attribute is used on a function that can cause the invoking
+thread to sleep.
+
+.. note::
+
+   This attribute is of relevance specifically when considering
+   applications that use only non-preemptible threads, because the
+   kernel will not replace a running cooperative-only thread at a
+   reschedule point unless that thread has explicitly invoked an
+   operation that caused it to sleep.
+
+   This attribute does not imply the function will sleep
+   unconditionally, but that the state of the system may require the
+   invoking thread to suspend, wait, or invoke :c:func:`k_yield()`
+   before it can complete its operation.  This behavior may be mediated
+   by **no-wait**.
+
+   Functions that are **sleep** are implicitly **reschedule**.
+
+   Functions that are **sleep** may be invoked from thread context.
+
+   Functions that are **sleep** may be invoked from interrupt and
+   pre-kernel contexts if and only if invoked in **no-wait** mode.
+
+.. _api_term_no-wait:
+
+no-wait
+=======
+
+The no-wait attribute is used on a function to indicate that a parameter
+to the function can force an execution path that will not cause the
+invoking thread to sleep.
+
+.. note::
+
+   The paradigmatic case of a no-wait function is a function that takes
+   a timeout, to which `K_NO_WAIT` can be passed.  The semantics of this
+   special timeout value are to execute the function's operation as long
+   as it can be completed immediately, and to return an error code
+   rather than sleep if it cannot.
+
+   It is use of the no-wait feature that allows functions like
+   :c:func:`k_sem_take` to be invoked from ISRs, since it is not
+   permitted to sleep in interrupt context.
+
+   **NOTE** That a function has a no-wait path does not imply that taking
```

Comment:
Not sure you need a NOTE within a note.  Also this sentence reads funny, maybe instead:
```
   A function with a no-wait path does not imply that taking
   that path guarantees the function is synchronous.
```